spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/SkipperStreamService.java

```diff
@@ -0,0 +1,259 @@
+		Release release = this.skipperStreamDeployer.upgradeStream(releaseName, packageIdentifier, yamlProperties);
+
+		if (release != null) {
+			updateStreamDefinitionFromReleaseManifest(streamName, release.getManifest());
+		}
+		else {
+			logger.warn("Missing release after Stream Update!");
+		}
+	}
+
+	private void updateStreamDefinitionFromReleaseManifest(String streamName, String releaseManifest) {
+
+		List<SpringCloudDeployerApplicationManifest> appManifests =
+				new SpringCloudDeployerApplicationManifestReader().read(releaseManifest);
+
+		Map<String, SpringCloudDeployerApplicationManifest> appManifestMap = new HashMap<>();
+
+		for (SpringCloudDeployerApplicationManifest am : appManifests) {
+			String name = am.getSpec().getApplicationProperties().get("spring.cloud.dataflow.stream.app.label");
+			appManifestMap.put(name, am);
+		}
+
+		StreamDefinition streamDefinition = this.streamDefinitionRepository.findOne(streamName);
+
+		LinkedList<StreamAppDefinition> updatedStreamAppDefinitions = new LinkedList<>();
+		for (StreamAppDefinition appDefinition : streamDefinition.getAppDefinitions()) {
+			StreamAppDefinition.Builder appDefinitionBuilder = StreamAppDefinition.Builder.from(appDefinition);
+			SpringCloudDeployerApplicationManifest applicationManifest = appManifestMap.get(appDefinition.getName());
+			// overrides app definition properties with those from the release manifest
+			appDefinitionBuilder.setProperties(applicationManifest.getSpec().getApplicationProperties());
+			updatedStreamAppDefinitions.addLast(appDefinitionBuilder.build(streamDefinition.getName()));
+
+			//Check and register any missing app version
+			updateAppVersionIfChanged(appDefinition, applicationManifest);
+		}
+
+		String dslText = new StreamDefinitionToDslConverter().toDsl(updatedStreamAppDefinitions);
+
+		StreamDefinition updatedStreamDefinition = new StreamDefinition(streamName, dslText);
+		logger.debug("Updated StreamDefinition: " + updatedStreamDefinition);
+
+		// TODO consider adding an explicit UPDATE method to the streamDefRepository
+		// Note: Not transactional and can lead to loosing the stream definition
+		this.streamDefinitionRepository.delete(updatedStreamDefinition);
+		this.streamDefinitionRepository.save(updatedStreamDefinition);
+	}
+
+	private void updateAppVersionIfChanged(StreamAppDefinition appDefinition,
+			SpringCloudDeployerApplicationManifest appManifest) {
+
+		String version = appManifest.getSpec().getVersion();
+		//String name = appManifest.getSpec().getApplicationProperties().get("spring.cloud.dataflow.stream.app.label");
+		String name = appDefinition.getRegisteredAppName();
+		ApplicationType type = ApplicationType.valueOf(
+				appManifest.getSpec().getApplicationProperties().get("spring.cloud.dataflow.stream.app.type"));
+		String resource = appManifest.getSpec().getApplicationProperties().get("spring.cloud.dataflow.stream.app.type");
+
+		if (!this.appRegistryService.appExist(name, type, version)) {
```

Comment:
We probably need some tests (maybe as part of this PR or as a separate PR) to check the flow here.

Reply:
You have a point i was thinking to add tests in the context of #1840, but i would try to add some unit tests for the logic here.